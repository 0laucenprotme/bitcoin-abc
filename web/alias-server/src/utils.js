'use strict';
const config = require('../config');

module.exports = {
    getValidAliasTxsToBeAddedToDb: function (validAliasesInDb, validAliasTxs) {
        /*
        - See if any tx exist in validAliasTxs and not in validAliasesinDb
        - Return validAliasTxsToBeAddedToDb, an array of these aliases
        - Note that validAliasesInDb and validAliasTxs are sorted by blockheight
        */
        const validAliasTxsToBeAddedToDbCount =
            validAliasTxs.length - validAliasesInDb.length;
        if (validAliasTxsToBeAddedToDbCount > 0) {
            const validAliasTxsToBeAddedToDb = validAliasTxs.slice(
                -validAliasTxsToBeAddedToDbCount,
            );
            return validAliasTxsToBeAddedToDb;
        } else {
            return [];
        }
    },
    getConfirmedTxsToBeAddedToDb: function (confirmedTxsInDb, confirmedTxs) {
        /*
        - See if any confirmed tx exists in confirmedTxs and not in confirmedTxsInDb
        - Return confirmedTxsToBeAddedToDb, an array of these txs
        - Note that confirmedTxsInDb and confirmedTxs are sorted by blockheight, highest to lowest;
          hence you want the first ${confirmedTxsToBeAddedToDbCount} entries of confirmedTxs
        */
        const confirmedTxsToBeAddedToDbCount =
            confirmedTxs.length - confirmedTxsInDb.length;
        if (confirmedTxsToBeAddedToDbCount > 0) {
            const confirmedTxsToBeAddedToDb = confirmedTxs.slice(
                0,
                confirmedTxsToBeAddedToDbCount,
            );
            return confirmedTxsToBeAddedToDb;
        } else {
            return [];
        }
    },
    generateReservedAliasTxArray: function () {
        const reservedAliasTxs = [];
        for (
            let i = 0;
            i < config.aliasConstants.reservedAliases.length;
            i += 1
        ) {
            reservedAliasTxs.push({
                address: config.aliasConstants.address,
                alias: config.aliasConstants.reservedAliases[i],
                blockheight: 1,
                txid: 'reserved',
            });
        }
        return reservedAliasTxs;
    },
    getAliasFromHex: function (aliasHex) {
        return Buffer.from(aliasHex, 'hex').toString('utf8');
    },
    getHexFromAlias: function (alias) {
        return Buffer.from(alias, 'utf8').toString('hex');
    },
    getAliasBytecount: function (alias) {
        const aliasHex = module.exports.getHexFromAlias(alias);
        const aliasByteCount = aliasHex.length / 2;
        return aliasByteCount;
    },
    isValidAliasString: function (alias) {
        /*
        Initial launch will support only lower case roman alphabet and numbers 0 through 9
        */
        return /^[a-z0-9]+$/.test(alias);
    },
    removeUnconfirmedTxsFromTxHistory: function (txHistory) {
        // Remove unconfirmed txs from an array of chronik tx objects
        const confirmedTxHistory = [];
        for (let i = 0; i < txHistory.length; i += 1) {
            const thisTx = txHistory[i];
            if (typeof thisTx.block !== 'undefined') {
                confirmedTxHistory.push(thisTx);
            }
        }
        return confirmedTxHistory;
    },
};
