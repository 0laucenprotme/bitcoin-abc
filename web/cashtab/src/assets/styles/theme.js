export const theme = {
    iconOutlined: '#3e3f42',
    formBorders: '#e7edf3',
    formAddonBackground: '#f4f4f4',
    formAddonForeground: '#3e3f42',
};
