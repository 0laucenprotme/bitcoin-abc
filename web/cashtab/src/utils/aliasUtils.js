import { currency } from 'components/Common/Ticker';

/*
 @response:
  [
    {alias: 'foo', address: 'ecash:qwert...'},
    {alias: 'foo2', address: 'ecash:qwert...'},
    {alias: 'foo3', address: 'ecash:qwert...'},
  ]
*/
export const getAliasServerHistory = async () => {
    let aliasServerResp, aliasServerRespJson;
    try {
        aliasServerResp = await fetch(
            currency.aliasSettings.aliasServerBaseUrl + '/aliases',
        );
        aliasServerRespJson = aliasServerResp.json();
    } catch (err) {
        console.log(
            `getAliasServerHistory(): Error retrieving aliases from alias-server`,
            err,
        );
    }

    return aliasServerRespJson;
};

/*
 @response:
  [
    {alias: 'foo', address: 'ecash:qwert...', txid: 'as12d1f324asdf'},
    {alias: 'foo2', address: 'ecash:qwert...' txid: 'as12d1f324asdf'},
    {alias: 'foo3', address: 'ecash:qwert...' txid: 'as12d1f324asdf'},
  ]
*/
export const getPendingAliases = async () => {
    let pendingAliasesResp, pendingAliasesRespJson;
    try {
        pendingAliasesResp = await fetch(
            currency.aliasSettings.aliasServerBaseUrl + '/pending',
        );
        pendingAliasesRespJson = pendingAliasesResp.json();
    } catch (err) {
        console.log(
            `getPendingAliases(): Error retrieving pending aliases from alias-server`,
            err,
        );
        return false;
    }

    return pendingAliasesRespJson;
};
