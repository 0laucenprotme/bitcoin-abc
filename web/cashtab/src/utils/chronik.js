// Chronik methods
import BigNumber from 'bignumber.js';
/* 
Note: chronik.script('p2pkh', hash160).utxos(); is not readily mockable in jest
Hence it is necessary to keep this out of any functions that require unit testing
*/
export const getUtxosSingleHashChronik = async (chronik, hash160) => {
    // Get utxos at a single address, which chronik takes in as a hash160
    let utxos;
    try {
        utxos = await chronik.script('p2pkh', hash160).utxos();
        if (utxos.length === 0) {
            // Chronik returns an empty array if there are no utxos at this hash160
            return [];
        }
        /* Chronik returns an array of with a single object if there are utxos at this hash 160
        [
            {
                outputScript: <hash160>,
                utxos:[{utxo}, {utxo}, ..., {utxo}]
            }
        ]
        */

        // Return only the array of utxos at this address
        return utxos[0].utxos;
    } catch (err) {
        console.log(`Error in chronik.utxos(${hash160})`);
        console.log(err);
    }
};

export const returnGetUtxosChronikPromise = (chronik, hash160AndAddressObj) => {
    /*
        Chronik thinks in hash160s, but people and wallets think in addresses
        Add the address to each utxo
    */
    return new Promise((resolve, reject) => {
        getUtxosSingleHashChronik(chronik, hash160AndAddressObj.hash160).then(
            result => {
                for (let i = 0; i < result.length; i += 1) {
                    const thisUtxo = result[i];
                    thisUtxo.address = hash160AndAddressObj.address;
                }
                resolve(result);
            },
            err => {
                reject(err);
            },
        );
    });
};

export const getUtxosChronik = async (chronik, hash160sMappedToAddresses) => {
    /* 
        Chronik only accepts utxo requests for one address at a time
        Construct an array of promises for each address
        Note: Chronik requires the hash160 of an address for this request
    */
    const chronikUtxoPromises = [];
    for (let i = 0; i < hash160sMappedToAddresses.length; i += 1) {
        const thisPromise = returnGetUtxosChronikPromise(
            chronik,
            hash160sMappedToAddresses[i],
        );
        chronikUtxoPromises.push(thisPromise);
    }
    const allUtxos = await Promise.all(chronikUtxoPromises);
    // Since each individual utxo has address information, no need to keep them in distinct arrays
    // Combine into one array of all utxos
    const flatUtxos = allUtxos.flat();
    return flatUtxos;
};

export const organizeUtxosByType = chronikUtxos => {
    /* 
    
    Convert chronik utxos (returned by getUtxosChronik function, above) to match 
    shape of existing slpBalancesAndUtxos object
    
    This means sequestering eToken utxos from non-eToken utxos

    For legacy reasons, the term "SLP" is still sometimes used to describe an eToken

    So, SLP utxos === eToken utxos, it's just a semantics difference here
    
    */

    const nonSlpUtxos = [];
    const preliminarySlpUtxos = [];
    for (let i = 0; i < chronikUtxos.length; i += 1) {
        // Construct nonSlpUtxos and slpUtxos arrays
        const thisUtxo = chronikUtxos[i];
        if (typeof thisUtxo.slpToken !== 'undefined') {
            preliminarySlpUtxos.push(thisUtxo);
        } else {
            nonSlpUtxos.push(thisUtxo);
        }
    }

    return { preliminarySlpUtxos, nonSlpUtxos };
};

export const getPreliminaryTokensArray = preliminarySlpUtxos => {
    // Iterate over the slpUtxos to create the 'tokens' object
    let tokensById = {};

    preliminarySlpUtxos.forEach(preliminarySlpUtxo => {
        /* 
        Note that a wallet could have many eToken utxos all belonging to the same eToken
        For example, a user could have 100 of a certain eToken, but this is composed of
        four utxos, one for 17, one for 50, one for 30, one for 3        
        */

        // Start with the existing object for this particular token, if it exists
        let token = tokensById[preliminarySlpUtxo.slpMeta.tokenId];

        if (token) {
            if (preliminarySlpUtxo.slpToken.amount) {
                token.balance = token.balance.plus(
                    new BigNumber(preliminarySlpUtxo.slpToken.amount),
                );
            }
        } else {
            // If it does not exist, create it
            token = {};
            token.tokenId = preliminarySlpUtxo.slpMeta.tokenId;
            if (preliminarySlpUtxo.slpToken.amount) {
                token.balance = new BigNumber(
                    preliminarySlpUtxo.slpToken.amount,
                );
            } else {
                token.balance = new BigNumber(0);
            }
            tokensById[preliminarySlpUtxo.slpMeta.tokenId] = token;
        }
    });

    const preliminaryTokensArray = Object.values(tokensById);
    return preliminaryTokensArray;
};

const returnGetTokenInfoChronikPromise = (chronik, tokenId) => {
    /*
    The chronik.tx(txid) API call returns extensive transaction information
    For the purposes of finalizing token information, we only need the token metadata

    This function returns a promise that extracts only this needed information from
    the chronik.tx(txid) API call

    In this way, calling Promise.all() on an array of tokenIds that lack metadata
    will return an array with all required metadata
    */
    return new Promise((resolve, reject) => {
        chronik.tx(tokenId).then(
            result => {
                if (typeof result === 'undefined') {
                    console.log(`result`, result);
                }

                const thisTokenInfo = result.slpTxData.genesisInfo;
                thisTokenInfo.tokenId = tokenId;
                // You only want the genesis info for tokenId
                resolve(thisTokenInfo);
            },
            err => {
                reject(err);
            },
        );
    });
};

export const processPreliminaryTokensArray = (
    preliminaryTokensArray,
    tokenInfoByTokenId,
) => {
    /* Iterate over preliminaryTokensArray to

    1 - Add slp metadata (token ticker, name, other metadata)
    2 - Calculate the token balance. Token balance in 
        preliminaryTokensArray does not take into account the
        decimal places of the token...so it is incorrect.

    */
    const finalTokenArray = [];
    for (let i = 0; i < preliminaryTokensArray.length; i += 1) {
        const thisToken = preliminaryTokensArray[i];
        const thisTokenId = thisToken.tokenId;

        // Because tokenInfoByTokenId is indexed by tokenId, it's easy to reference
        const thisTokenInfo = tokenInfoByTokenId[thisTokenId];

        // The decimals are specifically needed to calculate the correct balance
        const thisTokenDecimals = thisTokenInfo.decimals;

        // Add info object to token
        thisToken.info = thisTokenInfo;

        // Update balance according to decimals
        thisToken.balance = thisToken.balance.shiftedBy(-1 * thisTokenDecimals);

        // Now that you have the metadata and the correct balance,
        // preliminaryTokenInfo is finalTokenInfo
        finalTokenArray.push(thisToken);
    }
    return finalTokenArray;
};

export const finalizeTokensArray = async (
    chronik,
    preliminaryTokensArray,
    cachedTokenInfoById = {},
) => {
    console.log(
        `finalizeTokensArray called with cachedTokenInfoById`,
        cachedTokenInfoById,
    );
    // Iterate over preliminaryTokensArray to determine what tokens you need to make API calls for

    // Create an array of promises
    // Each promise is a chronik API call to obtain token metadata for this token ID
    const getTokenInfoPromises = [];

    for (let i = 0; i < preliminaryTokensArray.length; i += 1) {
        const thisTokenId = preliminaryTokensArray[i].tokenId;
        // See if you already have this info in cachedTokenInfo
        if (thisTokenId in cachedTokenInfoById) {
            // If you already have this info in cache, do not create an API request for it
            continue;
        }
        const thisTokenInfoPromise = returnGetTokenInfoChronikPromise(
            chronik,
            thisTokenId,
        );
        getTokenInfoPromises.push(thisTokenInfoPromise);
    }

    // For this test plan, to be removed later in stack
    console.log(
        `Cashtab asking chronik for token info about ${getTokenInfoPromises.length} tokens`,
    );
    console.log(`getTokenInfoPromises.length`, getTokenInfoPromises.length);

    const newTokensToCache = getTokenInfoPromises.length > 0;

    // Get all the token info you need
    let tokenInfoArray = [];
    try {
        tokenInfoArray = await Promise.all(getTokenInfoPromises);
    } catch (err) {
        console.log(`Error in Promise.all(getTokenInfoPromises)`, err);
    }

    // Add the token info you received from those API calls to
    // your token info cache object, cachedTokenInfoByTokenId

    const updatedTokenInfoById = cachedTokenInfoById;
    for (let i = 0; i < tokenInfoArray.length; i += 1) {
        /* tokenInfoArray is an array of objects that look like
        {
            "tokenTicker": "ST",
            "tokenName": "ST",
            "tokenDocumentUrl": "developer.bitcoin.com",
            "tokenDocumentHash": "",
            "decimals": 0,
            "tokenId": "bf24d955f59351e738ecd905966606a6837e478e1982943d724eab10caad82fd"
        }
        */

        const thisTokenInfo = tokenInfoArray[i];
        const thisTokenId = thisTokenInfo.tokenId;
        // Add this entry to updatedTokenInfoById
        updatedTokenInfoById[thisTokenId] = thisTokenInfo;
    }

    // Now use cachedTokenInfoByTokenId object to finalize token info
    // Split this out into a separate function so you can unit test
    const finalTokenArray = processPreliminaryTokensArray(
        preliminaryTokensArray,
        updatedTokenInfoById,
    );

    return { finalTokenArray, updatedTokenInfoById, newTokensToCache };
};

export const finalizeSlpUtxos = (preliminarySlpUtxos, tokenInfoById) => {
    // We need tokenQty in each slpUtxo to support transaction creation
    // Add this info here
    const finalizedSlpUtxos = [];
    for (let i = 0; i < preliminarySlpUtxos.length; i += 1) {
        const thisUtxo = preliminarySlpUtxos[i];
        const thisTokenId = thisUtxo.slpMeta.tokenId;
        const { decimals } = tokenInfoById[thisTokenId];
        // Update balance according to decimals
        thisUtxo.tokenQty = new BigNumber(thisUtxo.slpToken.amount)
            .shiftedBy(-1 * decimals)
            .toString();
        // SLP utxos also require tokenId and decimals directly in the utxo object
        // This is bad organization but necessary until bch-js is refactored
        // https://github.com/Permissionless-Software-Foundation/bch-js/blob/master/src/slp/tokentype1.js#L217
        thisUtxo.tokenId = thisTokenId;
        thisUtxo.decimals = decimals;
        finalizedSlpUtxos.push(thisUtxo);
    }
    return finalizedSlpUtxos;
};
