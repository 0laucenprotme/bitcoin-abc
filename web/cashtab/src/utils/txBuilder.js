// Transaction builder module streamlined for eCash
// reference: https://github.com/Permissionless-Software-Foundation/bch-js/blob/master/src/transaction-builder.js
const eCash = require('ecashjs-lib');
import coininfo from 'utils/coininfo';
const bip66 = require('bip66');
const bip68 = require('bc-bip68');
import cashaddr from 'ecashaddrjs';

class TransactionBuilder {
    static setAddress(address) {
        TransactionBuilder._address = address;
    }

    constructor() {
        const bitcoincashBitcoinJSLib = coininfo.bitcoincash.main.toBitcoinJS();
        this.transaction = new eCash.TransactionBuilder(
            bitcoincashBitcoinJSLib,
        );
        this.DEFAULT_SEQUENCE = 0xffffffff;
        this.hashTypes = {
            SIGHASH_ALL: 0x01,
            SIGHASH_NONE: 0x02,
            SIGHASH_SINGLE: 0x03,
            SIGHASH_ANYONECANPAY: 0x80,
            SIGHASH_BITCOINCASH_BIP143: 0x40,
            ADVANCED_TRANSACTION_MARKER: 0x00,
            ADVANCED_TRANSACTION_FLAG: 0x01,
        };
        this.signatureAlgorithms = {
            ECDSA: eCash.ECSignature.ECDSA,
            SCHNORR: eCash.ECSignature.SCHNORR,
        };
        this.bip66 = bip66;
        this.bip68 = bip68;
        this.p2shInput = false;
    }

    addInput(txHash, vout, sequence = this.DEFAULT_SEQUENCE, prevOutScript) {
        this.transaction.addInput(txHash, vout, sequence, prevOutScript);
    }

    addInputScript(vout, script) {
        this.tx = this.transaction.buildIncomplete();
        this.tx.setInputScript(vout, script);
        this.p2shInput = true;
    }

    addInputScripts(scripts) {
        this.tx = this.transaction.buildIncomplete();
        scripts.forEach(script => {
            this.tx.setInputScript(script.vout, script.script);
        });
        this.p2shInput = true;
    }

    // Customized addOutput() including streamlined toLegacyAddress() logic from bch-js
    addOutput(scriptPubKey, amount) {
        try {
            const decoded = cashaddr.decode(scriptPubKey);

            // toLegacyAddress logic from bch-js
            // reference: https://github.com/Permissionless-Software-Foundation/bch-js/blob/906d9b209404bf404ca4f12fdde2f32751e9635c/src/address.js#L57
            let version;
            switch (decoded.type) {
                case 'P2PKH':
                    version = coininfo.bitcoincash.main.versions.public;
                    break;
                case 'P2SH':
                    version = coininfo.bitcoincash.main.versions.scripthash;
                    break;
                default:
                    throw new Error(
                        `Unsupported address type : ${decoded.type}`,
                    );
            }
            const legacyAddress = eCash.address.toBase58Check(
                Buffer.from(decoded.hash),
                version,
            );

            this.transaction.addOutput(legacyAddress, amount);
        } catch (error) {
            // if output[0] was an OP_RETURN hex, add directly
            this.transaction.addOutput(scriptPubKey, amount);
        }
    }

    setLockTime(locktime) {
        this.transaction.setLockTime(locktime);
    }

    sign(
        vin,
        keyPair,
        redeemScript,
        hashType = this.hashTypes.SIGHASH_ALL,
        value,
        signatureAlgorithm,
    ) {
        let witnessScript;

        this.transaction.sign(
            vin,
            keyPair,
            redeemScript,
            hashType,
            value,
            witnessScript,
            signatureAlgorithm,
        );
    }

    build() {
        if (this.p2shInput === true) return this.tx;
        return this.transaction.build();
    }
}

export default TransactionBuilder;
