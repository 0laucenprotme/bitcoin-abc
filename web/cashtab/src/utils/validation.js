import { currency } from '@components/Common/Ticker.js';

// Validate cash amount
export const shouldRejectAmountInput = (
    cashAmount,
    selectedCurrency,
    fiatPrice,
    totalCashBalance,
) => {
    // Take cashAmount as input, a string from form input
    let error = false;
    let testedAmount = cashAmount;

    if (selectedCurrency === 'USD') {
        testedAmount = (cashAmount / fiatPrice).toFixed(8);
    }

    // Validate value for > 0
    if (isNaN(testedAmount)) {
        error = 'Amount must be a number';
    } else if (testedAmount <= 0) {
        error = 'Amount must be greater than 0';
    } else if (testedAmount < currency.dust) {
        error = `Send amount must be at least ${currency.dust} ${currency.ticker}`;
    } else if (testedAmount > totalCashBalance) {
        error = `Amount cannot exceed your ${currency.ticker} balance`;
    } else if (!isNaN(testedAmount) && testedAmount.toString().includes('.')) {
        if (testedAmount.toString().split('.')[1].length > 8) {
            error = `${currency.ticker} transactions do not support more than 8 decimal places`;
        }
    }
    // return false if no error, or string error msg if error
    return error;
};
