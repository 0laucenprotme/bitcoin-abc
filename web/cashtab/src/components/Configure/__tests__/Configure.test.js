import React from 'react';
import renderer from 'react-test-renderer';
import Configure from '../Configure';
import { ThemeProvider } from 'styled-components';
import { theme } from 'assets/styles/theme';
import { WalletContext } from 'utils/context';

test('Configure without a wallet', () => {
    const component = renderer.create(
        <WalletContext.Provider value={{ wallet: undefined }}>
            <ThemeProvider theme={theme}>
                <Configure />
            </ThemeProvider>
        </WalletContext.Provider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Configure with a wallet', () => {
    const component = renderer.create(
        <WalletContext.Provider
            value={{ wallet: { mnemonic: 'test mnemonic' } }}
        >
            <ThemeProvider theme={theme}>
                <Configure />
            </ThemeProvider>
        </WalletContext.Provider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
