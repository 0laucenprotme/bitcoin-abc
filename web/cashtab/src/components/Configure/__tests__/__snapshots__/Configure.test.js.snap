// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure with a wallet 1`] = `
<div
  className="ant-spin-nested-loading"
>
  <div>
    <div
      className="ant-spin ant-spin-spinning sc-bdVaJa iSePjH"
    >
      <span
        aria-label="loading"
        className="anticon anticon-loading anticon-spin ant-spin-dot"
        role="img"
        style={
          Object {
            "color": "#ff8d00",
            "fontSize": 24,
          }
        }
      >
        <svg
          aria-hidden="true"
          data-icon="loading"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="ant-spin-container ant-spin-blur"
  >
    <div
      className="sc-gZMcBi gcATnQ"
    >
      <h2>
        <i
          aria-label="icon: copy"
          className="anticon anticon-copy"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="copy"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
            />
          </svg>
        </i>
         Seed Phrase
      </h2>
      <p>
        Your seed phrase can be used to restore your wallet in case the original instance of it is destroyed. We highly recommend always making a copy of your seed phrase and keeping it somewhere safe.
      </p>
      <div
        className="ant-collapse ant-collapse-icon-position-left sc-bwzfXH dvdYjx"
        role={null}
      >
        <div
          className="ant-collapse-item"
        >
          <div
            aria-expanded="false"
            className="ant-collapse-header"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <i
              aria-label="icon: right"
              className="anticon anticon-right ant-collapse-arrow"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                />
              </svg>
            </i>
            Click to reveal seed phrase
          </div>
        </div>
      </div>
      <div
        className="sc-gqjmRU dWElEI"
      />
      <button
        className="sc-htpNat cPiyGy"
        onClick={[Function]}
      >
        <i
          aria-label="icon: plus-square"
          className="anticon anticon-plus-square"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="plus-square"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
            />
            <path
              d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
            />
          </svg>
        </i>
         New Wallet
      </button>
      <button
        className="sc-bxivhb bMNLGT"
        onClick={[Function]}
      >
        <i
          aria-label="icon: import"
          className="anticon anticon-import"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="import"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M888.3 757.4h-53.8c-4.2 0-7.7 3.5-7.7 7.7v61.8H197.1V197.1h629.8v61.8c0 4.2 3.5 7.7 7.7 7.7h53.8c4.2 0 7.7-3.4 7.7-7.7V158.7c0-17-13.7-30.7-30.7-30.7H158.7c-17 0-30.7 13.7-30.7 30.7v706.6c0 17 13.7 30.7 30.7 30.7h706.6c17 0 30.7-13.7 30.7-30.7V765.1c0-4.3-3.5-7.7-7.7-7.7zM902 476H588v-76c0-6.7-7.8-10.5-13-6.3l-141.9 112a8 8 0 0 0 0 12.6l141.9 112c5.3 4.2 13 .4 13-6.3v-76h314c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </i>
         Import Wallet
      </button>
    </div>
  </div>
</div>
`;

exports[`Configure without a wallet 1`] = `
<div
  className="ant-spin-nested-loading"
>
  <div>
    <div
      className="ant-spin ant-spin-spinning sc-bdVaJa iSePjH"
    >
      <span
        aria-label="loading"
        className="anticon anticon-loading anticon-spin ant-spin-dot"
        role="img"
        style={
          Object {
            "color": "#ff8d00",
            "fontSize": 24,
          }
        }
      >
        <svg
          aria-hidden="true"
          data-icon="loading"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="ant-spin-container ant-spin-blur"
  >
    <div
      className="sc-gZMcBi gcATnQ"
    >
      <h2>
        <i
          aria-label="icon: copy"
          className="anticon anticon-copy"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="copy"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
            />
          </svg>
        </i>
         Seed Phrase
      </h2>
      <p>
        Your seed phrase can be used to restore your wallet in case the original instance of it is destroyed. We highly recommend always making a copy of your seed phrase and keeping it somewhere safe.
      </p>
      <div
        className="sc-gqjmRU dWElEI"
      />
      <button
        className="sc-htpNat cPiyGy"
        onClick={[Function]}
      >
        <i
          aria-label="icon: plus-square"
          className="anticon anticon-plus-square"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="plus-square"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
            />
            <path
              d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
            />
          </svg>
        </i>
         New Wallet
      </button>
      <button
        className="sc-bxivhb bMNLGT"
        onClick={[Function]}
      >
        <i
          aria-label="icon: import"
          className="anticon anticon-import"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="import"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M888.3 757.4h-53.8c-4.2 0-7.7 3.5-7.7 7.7v61.8H197.1V197.1h629.8v61.8c0 4.2 3.5 7.7 7.7 7.7h53.8c4.2 0 7.7-3.4 7.7-7.7V158.7c0-17-13.7-30.7-30.7-30.7H158.7c-17 0-30.7 13.7-30.7 30.7v706.6c0 17 13.7 30.7 30.7 30.7h706.6c17 0 30.7-13.7 30.7-30.7V765.1c0-4.3-3.5-7.7-7.7-7.7zM902 476H588v-76c0-6.7-7.8-10.5-13-6.3l-141.9 112a8 8 0 0 0 0 12.6l141.9 112c5.3 4.2 13 .4 13-6.3v-76h314c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </i>
         Import Wallet
      </button>
    </div>
  </div>
</div>
`;
